File: Version Python/Ejercicio5.py
import tkinter as tk
from tkinter import messagebox

# Clase base
class Animal:
    def hacer_sonido(self):
        print("El animal hace un sonido.")
        return "El animal hace un sonido."

# Clase intermedia
class Mamifero(Animal):

    def alimentar(self):
        print("El mamífero se está alimentando.")
        return "El mamífero se está alimentando."

# Clase derivada
class Perro(Mamifero):

    def hacer_sonido(self):
        print("El perro hace el sonido: ¡Guau!")
        return "El perro hace el sonido: ¡Guau!"

# Interfaz gráfica
class Aplicacion:

    def __init__(self, ventana):

        self.ventana = ventana

        # Configuración de la ventana
        self.ventana.title(
            "Ejercicio 5 - Herencia Multinivel"
        )

        self.ventana.geometry("600x650")
        self.ventana.resizable(False, False)

        # Se crea una instancia de Perro, Perro hereda de Mamifero y Mamifero hereda de Animal.
        self.perro = Perro()

        # Título
        titulo = tk.Label(
            ventana,
            text="EJERCICIO 5: HERENCIA MULTINIVEL",
            font=("Arial", 18, "bold")
        )

        titulo.pack(pady=20)

        subtitulo = tk.Label(
            ventana,
            text="Herencia y sobrescritura de métodos",
            font=("Arial", 12)
        )

        subtitulo.pack()

        tk.Label(
            ventana,
            text="-" * 60
        ).pack(pady=10)

        # Jerarquía de clases
        tk.Label(
            ventana,
            text="Jerarquía de clases:",
            font=("Arial", 12, "bold")
        ).pack(pady=5)

        tk.Label(
            ventana,
            text=(
                "Animal\n"
                "   ↓\n"
                "Mamifero\n"
                "   ↓\n"
                "Perro"
            ),
            font=("Arial", 11),
            justify="center"
        ).pack(pady=5)

        # Botón "hacer sonido"
        tk.Button(
            ventana,
            text="Hacer sonido",
            width=30,
            command=self.hacer_sonido
        ).pack(pady=8)

        # Botón "alimentar"
        tk.Button(
            ventana,
            text="Alimentar",
            width=30,
            command=self.alimentar
        ).pack(pady=8)

        # Botón "demostrar herencia"
        tk.Button(
            ventana,
            text="Demostrar toda la herencia",
            width=30,
            command=self.demostrar_herencia
        ).pack(pady=8)

        # Área del ressultado
        tk.Label(
            ventana,
            text="Resultado:",
            font=("Arial", 12, "bold")
        ).pack(pady=(20, 5))

        self.resultado = tk.Label(
            ventana,
            text="Seleccione una opción.",
            font=("Arial", 11),
            justify="center",
            wraplength=520
        )

        self.resultado.pack(pady=10)

        # Botón "limpiar"
        tk.Button(
            ventana,
            text="Limpiar",
            width=30,
            command=self.limpiar
        ).pack(pady=5)

        # Botón "salir"
        tk.Button(
            ventana,
            text="Salir",
            width=30,
            command=ventana.destroy
        ).pack(pady=5)

    # Método "hacer sonido"
    def hacer_sonido(self):

        # Perro sobreescribe el método
        # hacer_sonido() definido originalmente
        # en Animal.
        resultado = self.perro.hacer_sonido()

        self.resultado.config(
            text=(
                "Método sobreescrito:\n\n"
                f"{resultado}\n\n"
                "Perro ha sobrescrito el método "
                "hacer_sonido() de Animal."
            )
        )

    # Método "alimentar"
    def alimentar(self):

        # Este método pertenece a Mamifero
        # y es heredado por Perro.
        resultado = self.perro.alimentar()

        self.resultado.config(
            text=(
                "Método heredado:\n\n"
                f"{resultado}\n\n"
                "Perro hereda el método alimentar() "
                "de Mamifero."
            )
        )

    # Demostración de la herencia
    def demostrar_herencia(self):

        sonido = self.perro.hacer_sonido()
        alimentacion = self.perro.alimentar()

        resultado = (
            "DEMOSTRACIÓN DE LA HERENCIA MULTINIVEL\n\n"
            "1. Animal\n"
            "   Define el método hacer_sonido().\n\n"
            "2. Mamifero\n"
            "   Hereda de Animal y añade alimentar().\n\n"
            "3. Perro\n"
            "   Hereda de Mamifero y sobrescribe "
            "hacer_sonido().\n\n"
            "RESULTADOS:\n\n"
            f"{sonido}\n"
            f"{alimentacion}"
        )

        self.resultado.config(
            text=resultado
        )

        messagebox.showinfo(
            "Demostración completada",
            "Se ejecutaron los métodos heredados "
            "y sobreescritos correctamente."
        )

    # Limpiar
    def limpiar(self):

        self.resultado.config(
            text="Seleccione una opción."
        )

# Inicio de aplicación

ventana = tk.Tk()

app = Aplicacion(ventana)

ventana.mainloop()

File: Version C#/Ejercicio5.cs
using System;

// Clase base
class Animal
{
    // Método virtual que puede ser sobrescrito por las clases derivadas
    public virtual void HacerSonido()
    {
        Console.WriteLine("El animal hace un sonido.");
    }
}

// Clase intermedia que hereda de Animal
class Mamifero : Animal
{
    // Método propio de Mamifero
    public void Alimentar()
    {
        Console.WriteLine("El mamífero se está alimentando.");
    }
}

// Clase derivada que hereda de Mamifero
class Perro : Mamifero
{
    // Sobrescritura del método HacerSonido() de Animal
    public override void HacerSonido()
    {
        Console.WriteLine("El perro hace el sonido ¡Guau!");
    }
}

class Program
{
    static void Main()
    {
        Console.Clear();

        Console.WriteLine("========================================");
        Console.WriteLine(" EJERCICIO 5: HERENCIA MULTINIVEL Y");
        Console.WriteLine("          SOBREESCRITURA");
        Console.WriteLine("========================================");

        // Se crea una instancia de Perro.
        // Perro hereda de Mamifero y Mamifero hereda de Animal.
        Perro perro = new Perro();

        int opcion;

        do
        {
            Console.WriteLine("\n----------------------------------------");
            Console.WriteLine("           MENÚ DEL EJERCICIO");
            Console.WriteLine("----------------------------------------");
            Console.WriteLine("1. Hacer sonido");
            Console.WriteLine("2. Alimentar");
            Console.WriteLine("3. Demostrar toda la herencia");
            Console.WriteLine("4. Salir");
            Console.Write("Seleccione una opción: ");

            // Valida que el usuario ingrese un número
            if (!int.TryParse(Console.ReadLine(), out opcion))
            {
                Console.WriteLine(
                    "\n[Error] Debe ingresar un número del 1 al 4."
                );

                continue;
            }

            switch (opcion)
            {
                case 1:
                    Console.WriteLine("\n[Método sobrescrito]");

                    // Se ejecuta la versión sobreescrita por Perro
                    perro.HacerSonido();

                    break;

                case 2:
                    Console.WriteLine("\n[Método heredado de Mamifero]");

                    // Perro hereda este método de la clase Mamifero
                    perro.Alimentar();

                    break;

                case 3:
                    Console.WriteLine(
                        "\n[Demostración de la herencia multinivel]"
                    );

                    Console.WriteLine(
                        "\n1. Animal -> define HacerSonido()"
                    );

                    Console.WriteLine(
                        "2. Mamifero -> hereda de Animal y añade Alimentar()"
                    );

                    Console.WriteLine(
                        "3. Perro -> hereda de Mamifero y sobreescribe HacerSonido()"
                    );

                    Console.WriteLine("\nEjecutando HacerSonido():");
                    perro.HacerSonido();

                    Console.WriteLine("\nEjecutando Alimentar():");
                    perro.Alimentar();

                    break;

                case 4:
                    Console.WriteLine(
                        "\nEl programa ha finalizado."
                    );

                    break;

                default:
                    Console.WriteLine(
                        "\n[Error] Opción no válida. " +
                        "Seleccione una opción del 1 al 4."
                    );

                    break;
            }

        } while (opcion != 4);

        Console.WriteLine("\nPresione ENTER para cerrar...");
        Console.ReadLine();
    }
}